service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java

```diff
@@ -94,99 +93,90 @@ public class EncounterComponentsMapper {
-
-        if (listReferencedToEncounter.isEmpty()) {
-            return StringUtils.EMPTY;
-        }
-
-        var entries = listReferencedToEncounter.orElseThrow().getEntry();
-        List<ListResource> topics = entries.stream()
-            .map(entry -> entry.getItem().getReferenceElement())
-            .map(reference ->
-                (ListResource) messageContext
-                    .getInputBundleHolder()
-                    .getRequiredResource(reference))
-            .collect(Collectors.toList());
+        return messageContext
+            .getInputBundleHolder()
+            .getListReferencedToEncounter(encounter.getIdElement(), CONSULTATION_LIST_CODE)
+            .map(this::mapListResourceToComponent)
+            .orElse(StringUtils.EMPTY);
+    }
 
-        return topics.stream()
+    private String mapListResourceToComponent(ListResource listResource) {
+        return listResource.getEntry()
+            .stream()
+            .map(EncounterComponentsMapper::getReferenceElement)
+            .map(this::getRequiredResource)
+            .map(ListResource.class::cast)
             .map(this::mapTopicListToComponent)
             .collect(Collectors.joining());
     }
 
     private String mapTopicListToComponent(ListResource topicList) {
-
         if (!CodeableConceptMappingUtils.hasCode(topicList.getCode(), List.of(TOPIC_LIST_CODE))) {
             throw new EhrMapperException(String.format("Unexpected list %s referenced in Consultation, expected list to be coded as "
                 + "Topic (EHR)", topicList.getId()));
         }
 
-        String components = mapTopicListComponents(topicList);
-
-        if (StringUtils.isAllEmpty(components)) {
-            return components;
-        }
-
-        String effectiveTime = prepareEffectiveTime(topicList);
-        String availabilityTime = prepareAvailabilityTime(topicList);
-
-        var params = CompoundStatementParameters.builder()
-            .nested(false)
-            .id(messageContext.getIdMapper().getOrNew(ResourceType.List, topicList.getIdElement()))
-            .classCode(TOPIC.getCode())
-            .compoundStatementCode(prepareCdForTopic(topicList))
-            .statusCode(COMPLETE_CODE)
-            .effectiveTime(effectiveTime)
-            .availabilityTime(availabilityTime)
-            .components(components)
-            .build();
-
-        return TemplateUtils.fillTemplate(COMPOUND_STATEMENT_TEMPLATE, params);
+        return buildCompoundStatement(
+            topicList,
+            TOPIC.getCode(),
+            prepareCdForTopic(topicList),
+            false,
+            mapTopicListComponents(topicList)
+        );
     }
 
     private String mapTopicListComponents(ListResource topicList) {
+        return String.join(
+            StringUtils.EMPTY,
+            mapCategorizedResources(topicList),
+            mapContainedResources(topicList),
+            mapUncategorizedResources(topicList)
```

Comment:
So from my understanding of this code, it's been introduced to handle resources which have been referenced in each of the two different forms:
```
List/abc#def
Observation/ghi
```

AFAIK is a low level implementation detail of how FHIR resources can live within the Bundle, and ideally if there is a method which takes in either of those Reference types, it should be able to return the `Resource` which it referrs to.

Furthermore, for this ticket it looks like we've fixed the situation where the resource entries are all of the form `Observation/xxx`, however if we had a mix and match of both forms of reference then the resources of the "contained" form won't be wrapped in the the Topic CompoundStatement.

I believe that our business logic should be agnostic of the type of Reference being used within the list, and instead should care more about the _type_ of Resource that the Reference is pointing at.

This method to me should have two cases:
1. I've been given a list of all topics, do X
2. I've been given a list of all non-list resources (irrespective of how they're referenced), do Y

Reply:
I've had a further look into this and the behaviour for "contained" resources includes the following requirement:
> The contained resource can’t be referred to outside the main resource, not even from another resource in the same bundle.

So therefore a resource in the form `List/abc#def` is invalid. The only valid way of referencing a contained resource is with the `#blah` form.